lib/autoupdate/tools/updater.go

```diff
@@ -0,0 +1,390 @@
+	requestedVersion := os.Getenv(teleportToolsVersionEnv)
+	switch {
+	// The user has turned off any form of automatic updates.
+	case requestedVersion == "off":
+		return "", false, nil
+	// Requested version already the same as client version.
+	case u.localVersion == requestedVersion:
+		return requestedVersion, false, nil
+	}
+
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		return "", false, trace.Wrap(err)
+	}
+	resp, err := webclient.Find(&webclient.Config{
+		Context:   ctx,
+		ProxyAddr: proxyAddr,
+		Pool:      certPool,
+		Timeout:   30 * time.Second,
+	})
+	if err != nil {
+		return "", false, trace.Wrap(err)
+	}
+
+	// If a version of client tools has already been downloaded to
+	// tools directory, return that.
+	toolsVersion, err := checkToolVersion(u.toolsDir)
+	if err != nil {
+		return "", false, trace.Wrap(err)
+	}
+
+	switch {
+	case requestedVersion != "" && requestedVersion != toolsVersion:
+		return requestedVersion, true, nil
+	case !resp.AutoUpdate.ToolsAutoUpdate || resp.AutoUpdate.ToolsVersion == "":
+		return "", false, nil
+	case u.localVersion == resp.AutoUpdate.ToolsVersion:
+		return resp.AutoUpdate.ToolsVersion, false, nil
+	case resp.AutoUpdate.ToolsVersion != toolsVersion:
+		return resp.AutoUpdate.ToolsVersion, true, nil
+	}
+
+	return toolsVersion, false, nil
+}
+
+// UpdateWithLock acquires filesystem lock, downloads requested version package, unarchive and replace existing one.
+func (u *Updater) UpdateWithLock(ctx context.Context, toolsVersion string) (err error) {
+	// Create tools directory if it does not exist.
+	if err := os.MkdirAll(u.toolsDir, 0o755); err != nil {
+		return trace.Wrap(err)
+	}
+	// Lock concurrent client tools execution util requested version is updated.
+	unlock, err := utils.FSWriteLock(filepath.Join(u.toolsDir, lockFileName))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		err = trace.NewAggregate(err, unlock())
+	}()
+
+	// If the version of the running binary or the version downloaded to
+	// tools directory is the same as the requested version of client tools,
+	// nothing to be done, exit early.
+	teleportVersion, err := checkToolVersion(u.toolsDir)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+
+	}
+	if toolsVersion == u.localVersion || toolsVersion == teleportVersion {
+		return nil
+	}
+
+	// Download and update client tools in tools directory.
+	if err := u.Update(ctx, toolsVersion); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return
+}
+
+// Update downloads requested version and replace it with existing one and cleanups the previous downloads
+// with defined updater directory suffix.
+func (u *Updater) Update(ctx context.Context, toolsVersion string) error {
+	// Get platform specific download URLs.
+	packages, err := teleportPackageURLs(u.baseUrl, toolsVersion)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	signalCtx, cancel := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
```

Comment:
Will the original context get canceled anyways if we receive these signals?

tsh/tctl/tbot will have their own signal handling, so it seems better to drive this via the original context.

Reply:
as far as I remember there were cases where we have no context propagated. I removed this part from library updater logic, will handle with tctl and tsh integration